apps/server/src/modules/authorization/authorization.service.ts

```diff
@@ -1,90 +1,82 @@
-		private readonly taskRule: TaskRule,
-		private readonly taskCardRule: TaskCardRule,
-		private readonly userRule: UserRule,
-		private readonly teamRule: TeamRule,
-		private readonly submissionRule: SubmissionRule,
+		private readonly ruleManager: RuleManager,
 		private readonly loader: ReferenceLoader,
-		private readonly schoolExternalToolRule: SchoolExternalToolRule
-	) {
-		super();
-		this.registerPermissions([
-			this.courseRule,
-			this.courseGroupRule,
-			this.lessonRule,
-			this.taskRule,
-			this.taskCardRule,
-			this.teamRule,
-			this.userRule,
-			this.schoolRule,
-			this.submissionRule,
-			this.schoolExternalToolRule,
-		]);
+		private readonly authorizationHelper: AuthorizationHelper
+	) {}
+
+	public checkIfAuthorized(user: User, entity: AuthorizableObject, context: AuthorizationContext): void {
+		if (!this.ruleManager.isAuthorized(user, entity, context)) {
+			throw new ForbiddenException();
+		}
+	}
+
+	public isAuthorized(user: User, entity: AuthorizableObject, context: AuthorizationContext): boolean {
+		return this.ruleManager.isAuthorized(user, entity, context);
 	}
 
-	checkPermission(user: User, entity: PermissionTypes, context: IPermissionContext) {
-		if (!this.hasPermission(user, entity, context)) {
+	public async checkIfAuthorizedByReferences(
+		userId: EntityId,
+		entityName: AllowedAuthorizationEntityType,
+		entityId: EntityId,
+		context: AuthorizationContext
+	): Promise<void> {
+		if (!(await this.isAuthorizedByReferences(userId, entityName, entityId, context))) {
 			throw new ForbiddenException();
 		}
 	}
 
-	async hasPermissionByReferences(
+	public async isAuthorizedByReferences(
 		userId: EntityId,
 		entityName: AllowedAuthorizationEntityType,
 		entityId: EntityId,
-		context: IPermissionContext
+		context: AuthorizationContext
 	): Promise<boolean> {
+		// TODO: I think this try-catch is unnecessary and wrong because there can be different reasons why the entity cannot be loaded and they should bubble up.
 		try {
 			const [user, entity] = await Promise.all([
 				this.getUserWithPermissions(userId),
 				this.loader.loadEntity(entityName, entityId),
 			]);
-			const permission = this.hasPermission(user, entity, context);
+			const isAuthorized = this.ruleManager.isAuthorized(user, entity, context);
 
-			return permission;
+			return isAuthorized;
 		} catch (err) {
 			throw new ForbiddenException(err);
 		}
 	}
 
-	async checkPermissionByReferences(
-		userId: EntityId,
-		entityName: AllowedAuthorizationEntityType,
-		entityId: EntityId,
-		context: IPermissionContext
-	) {
-		if (!(await this.hasPermissionByReferences(userId, entityName, entityId, context))) {
-			throw new ForbiddenException();
+	public checkAllPermissions(user: User, requiredPermissions: string[]): void {
+		if (!this.authorizationHelper.hasAllPermissions(user, requiredPermissions)) {
+			// TODO: Should be ForbiddenException
+			throw new UnauthorizedException();
 		}
 	}
```

Comment:
Just one quick question, when going through the code, I stumbled upon some instances of a `TODO` like

```js
public checkAllPermissions(user: User, requiredPermissions: string[]): void {
  if (!this.authorizationHelper.hasAllPermissions(user, requiredPermissions)) {
    // TODO: Should be ForbiddenException
    throw new UnauthorizedException();
  }
}
```
Why not change the exception to be of the desired type, instead of keeping the `TODO`? Or does anything speak against it?

Reply:
If I change the exceptions, the changes of this PR would not be a refactoring anymore, because it would change the beavior. E.g. we would need to check the client then, how it handles the exceptions. I want to make a separate ticket to change the exceptions.